crates/wasmparser/src/readers/core/elements.rs

```diff
@@ -49,49 +49,96 @@ pub enum ElementKind<'a> {
 }
 
 /// Represents the items of an element segment.
-#[derive(Debug, Copy, Clone)]
-pub struct ElementItems<'a> {
-    exprs: bool,
-    offset: usize,
-    data: &'a [u8],
+#[derive(Clone)]
+pub enum ElementItems<'a> {
+    /// This element contains function indices.
+    Functions(ElementFunctionItemsReader<'a>),
+    /// This element contains constant expressions used to initialize the table.
+    Expressions(ElementExpressionItemsReader<'a>),
 }
 
-/// Represents an individual item of an element segment.
-#[derive(Debug)]
-pub enum ElementItem<'a> {
-    /// The item is a function index.
-    Func(u32),
-    /// The item is an initialization expression.
-    Expr(ConstExpr<'a>),
+/// A reader for element function items in an element segment.
+#[derive(Clone)]
+pub struct ElementFunctionItemsReader<'a> {
+    reader: BinaryReader<'a>,
+    count: u32,
 }
 
-impl<'a> ElementItems<'a> {
-    /// Gets an items reader for the items in an element segment.
-    pub fn get_items_reader<'b>(&self) -> Result<ElementItemsReader<'b>>
-    where
-        'a: 'b,
-    {
-        ElementItemsReader::new(self.data, self.offset, self.exprs)
+impl<'a> ElementFunctionItemsReader<'a> {
+    /// Constructs a new `ElementFunctionItemsReader` for the given data and offset.
+    pub fn new(data: &'a [u8], offset: usize) -> Result<Self> {
+        let mut reader = BinaryReader::new_with_offset(data, offset);
+        let count = reader.read_var_u32()?;
+        Ok(Self { reader, count })
+    }
+
+    /// Gets the original position of the reader.
+    pub fn original_position(&self) -> usize {
+        self.reader.original_position()
+    }
+
+    /// Gets the count of element items in the segment.
+    pub fn get_count(&self) -> u32 {
+        self.count
+    }
+
+    /// Reads an element function item from the segment.
+    pub fn read(&mut self) -> Result<u32> {
+        let idx = self.reader.read_var_u32()?;
+        Ok(idx)
+    }
+}
+
+impl<'a> IntoIterator for ElementFunctionItemsReader<'a> {
```

Comment:
Could this perhaps use `SectionReader`, `SectionWithLimitedItems`, and `SectionIterator` to avoid a new iterator type? Albeit the name isn't great but I think could help cut down on the boilerplate possibly.

Reply:
Addressed!